### ISSUE TITLE
Image output of `NVDS` is invalid

### ISSUE THREAD

**Issue by jparismorgan** [RaymondWang987/NVDS#28]
Hi there, I'm trying to run `infer_NVDS_dpt_bi.py` on some images with `(NVDS) ~/repo/NVDS python infer_NVDS_dpt_bi.py --base_dir ./demo_outputs/dpt_init/000423/ --vnum 000423 --infer_w 896 --infer_h 384` using an M2 Macbook Pro (I changed the `device` to use `cpu`).

When I run, the normal depth network results look okay, i.e. `demo_outputs/dpt_init/000423/initial/gray/frame_000000.png`:

![frame_000000](https://github.com/RaymondWang987/NVDS/assets/1396242/b04423c1-7c20-4b11-a11a-92385a34583f)

But the results of the `NVDS` forward pass are all black, i.e. `demo_outputs/dpt_init/000423/1/gray/frame_000000.png`:

![frame_000000](https://github.com/RaymondWang987/NVDS/assets/1396242/430624ac-6b45-4a05-bcfe-2cb8ad6577ab)

and `demo_outputs/dpt_init/000423/1/color/0.png`:

![0](https://github.com/RaymondWang987/NVDS/assets/1396242/3b7143a9-1183-4319-b106-f8bfb4842801)

Curious if you have seen this before or have any suggestions to debug this? Thank you!

**Comment by jparismorgan** [comment]
In order to debug this more, and because it seems generally useful to others, I created a simple version of `infer_NVDS_dpt_bi.py` which should work with CUDA, MPS on Mac, or with CPU as a fallback. It just does the forward smoothing pass. Running it seemed to confirm something is not working with `NVDS`, as I get this from the first depth image created by `compute_depth()`:

![0](https://github.com/RaymondWang987/NVDS/assets/1396242/85c9118c-44db-4d1a-b1ab-bb069d158f52)

But I get this from the first depth image created by `smooth_depth()`:

![0](https://github.com/RaymondWang987/NVDS/assets/1396242/ef7bca3d-94c7-44d8-8151-d2ac611d732e)

Here is the code - perhaps you understand it better and will see that something is wrong with it? Or maybe you could run it on your machine and see if it works for you? Thank you for any help!!

```
import os
import argparse
import cv2
import numpy as np
import matplotlib.pyplot as plt
from backbone import *
import torch 
from networks import *
from full_model import *
import glob
from smooth_loss import *
from dpt.models import DPTDepthModel
from natsort import natsorted
import torchvision

# To get this to work on Mac with MPS I had to update to pytorch 2.1.0:
# - conda update pytorch
# - pip install chardet 
# Also added:
# - pip install natsort

torch.backends.cudnn.enabled = True
torch.backends.cudnn.benchmark = True

__mean = [0.485, 0.456, 0.406]
__std = [0.229, 0.224, 0.225]
__mean_dpt = [0.5, 0.5, 0.5]
__std_dpt = [0.5, 0.5, 0.5]

def get_args_parser():
    parser = argparse.ArgumentParser()

    parser.add_argument(
        "--input_dir",
        type=str
    )
    parser.add_argument(
        "--output_dir",
        type=str
    )
    parser.add_argument(
        "--infer_w",
        default= '896',
        type=int
    )
    parser.add_argument(
        "--infer_h",
        default= '384',
        type=int
    )

      # GMFlow model
    parser.add_argument('--num_scales', default=1, type=int,
                        help='basic gmflow model uses a single 1/8 feature, the refinement uses 1/4 feature')
    parser.add_argument('--feature_channels', default=128, type=int)
    parser.add_argument('--upsample_factor', default=8, type=int)
    parser.add_argument('--num_transformer_layers', default=6, type=int)
    parser.add_argument('--num_head', default=1, type=int)
    parser.add_argument('--attention_type', default='swin', type=str)
    parser.add_argument('--ffn_dim_expansion', default=4, type=int)

    # resume pretrained model or resume training
    parser.add_argument('--resume', default=None, type=str,
                        help='resume from pretrain model for finetuing or resume from terminated training')
    parser.add_argument('--strict_resume', action='store_true')
    parser.add_argument('--no_resume_optimizer', action='store_true')

    # inference on a directory
    #parser.add_argument('--inference_dir', default=None, type=str)
    parser.add_argument('--inference_size', default=None, type=int, nargs='+',
                        help='can specify the inference size')
    parser.add_argument('--dir_paired_data', action='store_true',
                        help='Paired data in a dir instead of a sequence')
    parser.add_argument('--save_flo_flow', action='store_true')
    parser.add_argument('--pred_bidir_flow', action='store_true',
                        help='predict bidirectional flow')
    parser.add_argument('--fwd_bwd_consistency_check', action='store_true',
                        help='forward backward consistency check with bidirection flow')

    parser.add_argument('--attn_splits_list', default=[2], type=int, nargs='+',
                        help='number of splits in attention')
    parser.add_argument('--corr_radius_list', default=[-1], type=int, nargs='+',
                        help='correlation radius for matching, -1 indicates global matching')
    parser.add_argument('--prop_radius_list', default=[-1], type=int, nargs='+',
                        help='self-attention radius for flow propagation, -1 indicates global attention')

    return parser

def img_loader(path):
    image = cv2.imread(path)
    if image.ndim == 2:
        image = cv2.cvtColor(image, cv2.COLOR_GRAY2BGR)
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB) / 255.0
    return image

def gt_png_loader(path):
    depth = cv2.imread(path, -1) #/ 65535.0
    return depth.astype(np.float32)

def get_depth_model(device):
  dpt = DPTDepthModel(
            path='./dpt/checkpoints/dpt_large-midas-2f21e586.pt',
            backbone="vitl16_384",
            non_negative=True,
            enable_attention_hooks=False,
        ).to(device)
  dpt.eval()
  return dpt

def get_nvds_model(device):
    checkpoint = torch.load('./NVDS_checkpoints/NVDS_Stabilizer.pth', map_location = 'cpu') 
    model = NVDS()
    model = torch.nn.DataParallel(model, device_ids=[0]) # .cuda()
    model.load_state_dict(checkpoint)
    model.to(device) 
    model.eval()
    return model

def compute_depth(image_paths, device, infer_size, save_disparity_dir, save_colorized_dir):
  os.makedirs(save_disparity_dir, exist_ok=True)
  os.makedirs(save_colorized_dir, exist_ok=True)

  dpt = get_depth_model(device)

  for i in range(len(image_paths)):
    frame = image_paths[i]
    rgb = img_loader(frame)
    rgb = cv2.resize(rgb, infer_size, interpolation=cv2.INTER_CUBIC)
    rgb = (rgb - __mean_dpt) / __std_dpt
    rgb = np.transpose(rgb, (2, 0, 1))
    rgb = torch.Tensor(np.ascontiguousarray(rgb).astype(np.float32)).unsqueeze(0)
    rgb = rgb.to(device)
    with torch.no_grad():
        outputs = dpt.forward(rgb)
        print(f"[{i}][{frame}]")
        print(f"  min(outputs): {torch.min(outputs)}, mean(outputs): {torch.mean(outputs)}, max(outputs): {torch.max(outputs)}")
        plt.imsave(
            save_colorized_dir+str(i)+'.png',
            outputs.cpu().numpy().squeeze(), 
            cmap='Spectral_r',
            vmin =np.min(outputs.cpu().numpy().squeeze()), 
            vmax = np.max(outputs.cpu().numpy().squeeze()))
        outputs = outputs.cpu().numpy().squeeze()
        depth_min = outputs.min()
        depth_max = outputs.max()
        outputs = 65535.0 * (outputs - depth_min) / (depth_max - depth_min)
        cv2.imwrite(
            save_disparity_dir+str(i)+'.png',
            outputs.astype("uint16"), 
            [cv2.IMWRITE_PNG_COMPRESSION, 0])

def smooth_depth(image_paths, depth_paths, device, infer_size, save_disparity_dir, save_colorized_dir, save_test_dir):
  os.makedirs(save_disparity_dir, exist_ok=True)
  os.makedirs(save_colorized_dir, exist_ok=True)
  os.makedirs(save_test_dir, exist_ok=True)
  
  model = get_nvds_model(device)
  seq_len = 4

  for i in range(len(image_paths)):
    frame = image_paths[i]
    depth_frame = depth_paths[i]
    print(f"[{i}][{frame}][{depth_frame}]")

    # Get the RGB image.
    img = img_loader(frame)
    img = cv2.resize(img, infer_size, interpolation=cv2.INTER_CUBIC)
    img = (img - __mean) / __std
    img = np.transpose(img, (2, 0, 1))
    img = torch.Tensor(np.ascontiguousarray(img).astype(np.float32))
    img = img.unsqueeze(0)
    torchvision.utils.save_image(img, save_test_dir+str(i)+'_img.png',)

    # Get the depth image.
    depth = gt_png_loader(depth_frame)
    depth = cv2.resize(depth, infer_size, interpolation=cv2.INTER_NEAREST)
    depth = torch.Tensor(np.ascontiguousarray(depth.astype(np.float32))).unsqueeze(0)
    depth = (depth-torch.min(depth))/(torch.max(depth)-torch.min(depth))
    depth = depth.unsqueeze(0)
    torchvision.utils.save_image(depth, save_test_dir+str(i)+'_depth.png',)

    # Create a RGBD image.
    rgbd = torch.cat([img, depth],dim=1)
    torchvision.utils.save_image(rgbd, save_test_dir+str(i)+'_rgbd.png',)

    # Set up the input to the NVDS model.
    for j in range(seq_len):
        if j == 0:
            ref_seq = rgbd
        else:
            ref_seq = torch.cat([ref_seq,rgbd],dim=0)
    ref_seq = ref_seq.unsqueeze(0)
    ref_seq = ref_seq.to(device)
    print(f"  Using ref_seq.shape {ref_seq.shape}")

    # Run the NVDS smoothing model.
    with torch.no_grad():
        outputs = model(ref_seq)
        print(f"  min(outputs): {torch.min(outputs)}, mean(outputs): {torch.mean(outputs)}, max(outputs): {torch.max(outputs)}")
        outputs = outputs.squeeze(1)

        plt.imsave(
            save_colorized_dir+str(i)+'.png',
            outputs.cpu().numpy().squeeze(), 
            cmap='Spectral_r',
            vmin =np.min(outputs.cpu().numpy().squeeze()),
            vmax = np.max(outputs.cpu().numpy().squeeze()))
        
        outputs = outputs.cpu().numpy().squeeze()
        depth_min = outputs.min()
        depth_max = outputs.max()
        outputs = 65535.0 * (outputs - depth_min) / (depth_max - depth_min)
        cv2.imwrite(
            save_disparity_dir+str(i)+'.png',
            outputs.astype("uint16"),
            [cv2.IMWRITE_PNG_COMPRESSION, 0])

def main(args):
    os.makedirs(args.output_dir, exist_ok=True)
    
    if torch.cuda.is_available():
        device = torch.device("cuda:0")
    elif torch.backends.mps.is_available() and torch.backends.mps.is_built():
        device = torch.device("mps:0")
    else:
        device = torch.device("cpu")
    infer_size = (int(args.infer_w),int(args.infer_h))
    print(f"Using device: {device}, infer_size: {device}")

    print('Computing depth maps.')
    frames = natsorted(glob.glob(args.input_dir + '/*.png'))
    save_disparity_dir=args.output_dir + '/0/gray/'
    compute_depth(
       image_paths=frames,
       device=device,
       infer_size=infer_size,
       save_disparity_dir=save_disparity_dir,
       save_colorized_dir=args.output_dir + '/0/color/')
    
    print('Smoothing with a forward pass.')
    depth_frames = natsorted(glob.glob(save_disparity_dir + '/*.png'))
    smooth_depth(
       image_paths=frames,
       depth_paths=depth_frames,
       device=device,
       infer_size=infer_size,
       save_disparity_dir=args.output_dir + '/1/gray/',
       save_colorized_dir=args.output_dir + '/1/color/',
       save_test_dir=args.output_dir + '/1/test/'
    )


if __name__ == '__main__':
  print('Starting.')
  parser = get_args_parser()
  args = parser.parse_args()
  main(args)
  print('Done.')
```

**Comment by luuude** [comment]
I ran your code, I had to alter some small things to get it to run on CUDA. Looks ok to me. But is it a good idea to store it as 16bit png? Would it not be better to user 32bit float in a exr file? I uploaded the results here https://ludvig-betong.filemail.com/d/agdtxtcjrtwszao

**Comment by jparismorgan** [comment]
Thanks for running it! I don't know about 16 vs 32 bit, I was just copying code existing in the repo already. But could be nice to change to 32. Would you mind sharing your code with the modifications? Maybe something that you had to fix for CUDA will also fix my issue. Thanks!

**Comment by luuude** [comment]
If I remember correctly I commented out :

    #elif torch.backends.mps.is_available() and torch.backends.mps.is_built():
        #device = torch.device("mps:0")

And maybe added ("cuda:0") somewhere.

Regarding the 16/32 float to int issue I think you loose some gradients when converting back and forth. So I imagine it would be best to just keep it in float all the way. I think you can store float in PNGs but maybe you have to modify it slightly, OpenEXR image files handles float without any modifications. Easiest way to use exr is with OpenImageIO.

**Comment by jparismorgan** [comment]
Thanks @luuude. I did find out the potential cause of this bug, which is that if I use `device = torch.device("cpu")` instead of `device = torch.device("mps:0")`, then I can run the `NVDS` model fine (and the script as a whole works), though it is very slow. So my current guesses of the error:
1. Somewhere in the codebase there is a bug and that bug makes MPS acceleration not work
2. There is a bug in Pytorch and we are hitting it in this code path

Not quite sure which it is, but if anyone has suggestions to debug, would love to hear!

**Comment by bigipalabom** [comment]
> Thanks @luuude. I did find out the potential cause of this bug, which is that if I use `device = torch.device("cpu")` instead of `device = torch.device("mps:0")`, then I can run the `NVDS` model fine (and the script as a whole works), though it is very slow. So my current guesses of the error:
> 
> 1. Somewhere in the codebase there is a bug and that bug makes MPS acceleration not work
> 2. There is a bug in Pytorch and we are hitting it in this code path
> 
> Not quite sure which it is, but if anyone has suggestions to debug, would love to hear!

hi, i located the issue

https://github.com/huggingface/transformers/issues/22468

> @amyeroberts just FIY, thanks the clue from pytorch dev team, the problem can be solved if using contiguous() tensor here:
> 
> https://github.com/huggingface/transformers/blob/1670be4bdec19d5a8893f943bf78a8d9b3dc8911/src/transformers/models/glpn/modeling_glpn.py#L283
> 
> something like this: `hidden_states = self.intermediate_act_fn(hidden_states.contiguous())`
> 
> Not sure it's worth a PR, since it's not a solution of the root problem, but if so, please let me know, I'll create one

**Comment by jparismorgan** [comment]
Hi @bigipalabom, amazing, that is a great find, thank you! And beyond it being a great find, it looks like it worked 😁 In `backbone.py` I updated to `x.contiguous()` and now the `NVDS` smoothing model works:
```
    def forward(self, x, H, W):
        x = self.fc1(x)
        x = self.dwconv(x, H, W)
        # Before it was x = self.act(x)
        x = self.act(x.contiguous())
        x = self.drop(x)
        x = self.fc2(x)
        x = self.drop(x)
        return x
```
I don't know enough PyTorch to understand if there are potential issues with this change. But cc @RaymondWang987, this seems like a nice update to make if it seems safe? It will unblock other Mac users who want to use MPS acceleration.

**Comment by velaia** [comment]
Great work @jparismorgan @bigipalabom ! I've tried to understand the issue a little deeper: the .contiguous() seems to be about the memory allocation of the tensors according to [this post](https://stackoverflow.com/a/69599806/258871). `self.act` is probably `nn.GELU` in this case which falls back to the functional implementation of GELU. I've looked for the MPS implementation of GELU and found [Activation.mm](https://github.com/pytorch/pytorch/blob/2728c9137dee6b38126b5a3622b857a18432590f/aten/src/ATen/native/mps/operations/Activation.mm) copyrighted by Apple, so I guess I'm not totally on the wrong track.

In the [corresponding commits](https://github.com/pytorch/pytorch/commits/2728c9137dee6b38126b5a3622b857a18432590f/aten/src/ATen/native/mps/operations/Activation.mm) I checked for activity mentioning GELU and found a couple of people working on it in the past. Maybe one of you can help, @malfet @qqaatw @DenisVieriu97 ? My best guess is that there's an assumption in the MPS GELU implementation that the tensor memory is contiguous but that's usually not the case. Unfortunately my Objective-C is non-existent 😉

**Comment by bigipalabom** [comment]
> Hi @bigipalabom, amazing, that is a great find, thank you! And beyond it being a great find, it looks like it worked 😁 In `backbone.py` I updated to `x.contiguous()` and now the `NVDS` smoothing model works:
> 
> ```
>     def forward(self, x, H, W):
>         x = self.fc1(x)
>         x = self.dwconv(x, H, W)
>         # NOTE(paris): self.act is nn.GELU
>         # Before it was x = self.act(x)
>         x = self.act(x.contiguous())
>         x = self.drop(x)
>         x = self.fc2(x)
>         x = self.drop(x)
>         return x
> ```
> 
> I don't know enough PyTorch to understand if there are potential issues with this change. But cc @RaymondWang987, this seems like a nice update to make if it seems safe? It will unblock other Mac users who want to use MPS acceleration.

Haha, glad it hepled. I remember Apple has similar problem ever since ancient macintosh time, as per Apple, "Think different"
